### DISCUSSION TITLE
Are tag-less container images deleted?

### DISCUSSION THREAD

**Discussion by airtower-luna** [community/community#26716] [Packages]
<p>I’ve been trying out the GHCR with a <a href="https://github.com/airtower-luna/hello-ghcr">very simple container</a>, including build on Actions. So far I have the build always push to the <code>beta</code> tag. I noticed that after a second git push that triggered the container build and push to GHCR there is no trace left of the image that was previously tagged <code>beta</code> (i.e., no untagged version in the versions list).</p>
<p>Is this intentional? I’m not complaining about the behavior, I just want to know if I can rely on it to e.g. push a <code>nightly</code> tag or similar without worrying about cleaning up old versions. :wink:</p>

**Comment by clarkbw** [comment, marked as answer]
<p>There should be an untagged version left behind assuming the new build is different. If there was no difference then you’d see this behaviour.  We’re not deleting tag-less containers, they stick around until you delete them.</p>

**Comment by airtower-luna** [comment]
<p>Thanks for the explanation! Content-wise the containers are identical, just different labels, so I suppose that explains it. :slightly_smiling_face:</p>
<p>In that case I will have to look into cleanup for nightly builds and the like.</p>

**Comment by airtower-luna** [comment]
<p>I experimented with this a bit more and confirmed that tag-less images are still accessible if you have the digest. However, there seems to be no way to discover them: They don’t show up on the Github package page (only tagged versions there), and the <a href="https://docs.docker.com/registry/spec/api/">Docker HTTP API V2 documentation</a> doesn’t mention any way to find them either.</p>
<p>Is there any other way I’m missing?</p>

**Comment by clarkbw** [comment]
<p>Can you take a look at the versions page? This is part of the settings page.  I thought it had all the versions available and you could click into them.<br>
</p><aside class="onebox allowlistedgeneric">
  <header class="source">
      <img src="https://github.githubassets.com/favicons/favicon.svg" class="site-icon" width="32" height="32">
      <a href="https://github.com/users/airtower-luna/packages/container/hello-ghcr-meow/versions" target="_blank" rel="noopener">GitHub</a>
  </header>
  <article class="onebox-body">
    <img src="https://user-images.githubusercontent.com/2134/181099355-1057a8b7-8e4f-4d5f-add0-2b22e377c422.png" class="thumbnail onebox-avatar" width="500" height="500" srcset="https://aws1.discourse-cdn.com/github/optimized/2X/3/337b7cc06184984e148c075df20ef009e0e6ecaa_2_500x500.png, https://aws1.discourse-cdn.com/github/optimized/2X/3/337b7cc06184984e148c075df20ef009e0e6ecaa_2_750x750.png 1.5x, https://aws1.discourse-cdn.com/github/optimized/2X/3/337b7cc06184984e148c075df20ef009e0e6ecaa_2_1000x1000.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/3/337b7cc06184984e148c075df20ef009e0e6ecaa_2_10x10.png">

<h3><a href="https://github.com/users/airtower-luna/packages/container/hello-ghcr-meow/versions" target="_blank" rel="noopener">Build software better, together</a></h3>

<p>GitHub is where people build software. More than 50 million people use GitHub to discover, fork, and contribute to over 100 million projects.</p>


  </article>
  <div class="onebox-metadata">
    
    
  </div>
  <div style="clear: both"></div>
</aside>

**Comment by airtower-luna** [comment]
<p>Evidently it shows only tagged versions. Two digests I still have saved for that image, both formerly tagged <code>beta</code>:</p>
<pre><code class="lang-nohighlight">sha256:8090e11e86f6a5be8138cfc5026e170123497bfa1e72ec75a6cb831135bba36b
sha256:676d25701f6af4f8e3b958392c30e18af0e71a48fcf3ccd4ef3621ab6b041ba3
</code></pre>
<p>I just tested that I can still pull the images using those digests (yes), but they don’t show up on the versions page, or anywhere else that I’m aware of.</p>

**Comment by jcansdale** [comment]
<aside class="quote no-group" data-username="airtower-luna" data-post="6" data-topic="132977">
<div class="title">
<div class="quote-controls"></div>
 airtower-luna:</div>
<blockquote>
<p>I just tested that I can still pull the images using those digests (yes), but they don’t show up on the versions page, or anywhere else that I’m aware of.</p>
</blockquote>
</aside>
<p>This is also how it works on Docker Hub.</p>
<p>How would you want the un-tagged images to appear? Would you want to see the tags the image was previous tagged with?</p>

**Comment by airtower-luna** [comment]
<aside class="quote group-github-employees" data-username="jcansdale" data-post="7" data-topic="132977">
<div class="title">
<div class="quote-controls"></div>
 jcansdale:</div>
<blockquote>
<p>How would you want the un-tagged images to appear?</p>
</blockquote>
</aside>
<p>Preferably listed via an API call (rationale below), although on the versions page (maybe behind an “show untagged images” button or something) would be nice.</p>
<aside class="quote group-github-employees" data-username="jcansdale" data-post="7" data-topic="132977">
<div class="title">
<div class="quote-controls"></div>
 jcansdale:</div>
<blockquote>
<p>Would you want to see the tags the image was previous tagged with?</p>
</blockquote>
</aside>
<p>Can’t hurt, but I’d be more interested in the timestamp of when the image was pushed.</p>
<p>What I’m thinking of is cleaning up images that aren’t needed after a while, e.g.:</p>
<ul>
<li>Nightly (or even per-commit) builds that could be used for integration testing, or development on related projects.</li>
<li>Build environments: I’d like to prepare environments for use that with the Actions <code>jobs.&lt;job_id&gt;.container</code> keyword so dependencies don’t have to be installed during every build. That takes about 2 min each time, per distro tested in a matrix build (well, except Alpine which needs just a few seconds). To get updates of the underlying distro and dependencies I’d want to rebuild those environments frequently, e.g. during the first CI build of the day.</li>
</ul>
<p>The latter would accumulate in the order of gigabytes pretty quickly, and would hardly be useful for anyone else in old versions, so I’d like to clean up after myself. Basically a sort of <code>docker image prune</code>. :wink:</p>
<p>As long as packages storage is free for open source projects it would not hurt me to leave them around, but for people with private images it could quickly become a billing issue. And evidently it became an issue for Docker Hub, too, given their <a href="https://www.docker.com/blog/scaling-dockers-business-to-serve-millions-more-developers-storage/">recent announcement to delete unused images on free accounts after 6 months</a> (that’d do the trick for me, too, but maybe not for people who need to pay for storage). That post also says:</p>
<blockquote>
<p>In addition, Docker will also be providing tooling, in the form of a UI and APIs, that will allow users to more easily manage their images.</p>
</blockquote>
<p>I haven’t seen any specifics on that yet, but to me it sound like “tools to help you clean up”, so a compatible API on GHCR might be a thing to consider.</p>

**Comment by jcansdale** [comment]
<p><a class="mention" href="/airtower-luna">@airtower-luna</a> thanks for the link to the announcement.</p>
<blockquote>
<ul>
<li>
<strong>Example <span class="hashtag">#1:</span></strong>  Molly, a free Docker Hub user, pushed a tagged image  <code>molly/hello-world:v1</code> to Docker Hub on January 1, 2019. The image was never pulled since it was pushed. This tagged image will be considered inactive beginning November 1, 2020 when the new policy takes effect. The image and any tag pointing to hit, will be subject to deletion on November 1, 2020.</li>
</ul>
</blockquote>
<p>Wow, this is some pretty agressive cleanup of tagged images!</p>
<blockquote>
<ul>
<li>
<strong>Example <span class="hashtag">#2</span></strong> : Molly has another untagged image  <code>molly/myapp@sha256:c0ffee</code>  that was first pushed on January 1, 2018. This image was last pulled on August 1, 2020. This image will be considered an active image and will not be subject to deletion on November 1, 2020.</li>
</ul>
</blockquote>
<p>For untagged images this seems pretty reasonable.</p>
<blockquote>
<p>What I’m thinking of is cleaning up images that aren’t needed after a while</p>
</blockquote>
<p>So you’d want to know when they were pushed, whether they’re tagged and maybe when they where last downloaded?</p>

**Comment by airtower-luna** [comment]
<aside class="quote group-github-employees" data-username="jcansdale" data-post="9" data-topic="132977">
<div class="title">
<div class="quote-controls"></div>
 jcansdale:</div>
<blockquote>
<p>So you’d want to know when they were pushed, whether they’re tagged and maybe when they where last downloaded?</p>
</blockquote>
</aside>
<p>Yes, exactly. I guess the most flexible solution would be to include a list of tags for the image, if any. With that information I could easily set up e.g. a scheduled Actions workflow to clean up old images that are either untagged or tagged in a certain way. :slightly_smiling_face:</p>
<p>And I agree that the Docker Hub policy is rather aggressive for tagged images, in principle I’d like to keep release versions around indefinitely. On the other hand depending on the image content I might worry about people using such an old image anyway, at least without vulnerability scanning (which I’d really like to see in GHCR, but that’s a separate thing). I wouldn’t want to keep track of which of the included libraries might have gotten security patches in the last half year, and whether any of the issues might be exploitable in the context of the image. :sweat_smile:</p>

**Comment by airtower-luna** [comment]
<p>I’ve written the kind of cleanup script I was thinking of here using the new <a href="https://github.blog/changelog/2021-02-17-packages-delete-and-restore-for-all-packages/">packages delete/restore API</a>. :slightly_smiling_face:</p>
<aside class="onebox githubblob">
  <header class="source">
      <a href="https://github.com/airtower-luna/hello-ghcr/blob/e17ebc574711f4bcf85f8bd6b2380e226901778d/ghcr-prune.py#L76-L85" target="_blank" rel="noopener">github.com</a>
  </header>
  <article class="onebox-body">
    <h4><a href="https://github.com/airtower-luna/hello-ghcr/blob/e17ebc574711f4bcf85f8bd6b2380e226901778d/ghcr-prune.py#L76-L85" target="_blank" rel="noopener">airtower-luna/hello-ghcr/blob/e17ebc574711f4bcf85f8bd6b2380e226901778d/ghcr-prune.py#L76-L85</a></h4>
<pre class="onebox"><code class="lang-py"><ol class="start lines" start="76" style="counter-reset: li-counter 75 ;">
<li># prune old untagged images if requested</li>
<li>if del_before is not None and created &lt; del_before \</li>
<li>   and len(metadata['tags']) == 0:</li>
<li>    if args.dry_run:</li>
<li>        print(f'would delete {v["id"]}')</li>
<li>    else:</li>
<li>        r = s.delete(f'https://api.github.com/user/packages/'</li>
<li>                     f'container/{args.container}/versions/{v["id"]}')</li>
<li>        r.raise_for_status()</li>
<li>        print(f'deleted {v["id"]}')</li>
</ol></code></pre>


  </article>
  <div class="onebox-metadata">
    
    
  </div>
  <div style="clear: both"></div>
</aside>

**Comment by jcansdale** [comment]
<p>Hi <a class="mention" href="/airtower-luna">@airtower-luna</a>,</p>
<blockquote>
<p>I’ve written the kind of cleanup script I was thinking of here using the new <a href="https://github.blog/changelog/2021-02-17-packages-delete-and-restore-for-all-packages/" rel="noopener nofollow ugc">packages delete/restore API </a>. :slightly_smiling_face:</p>
</blockquote>
<p>I had a similar thought. :smiley:</p>
<p>Here’s a GitHub Action script that can be used to delete untagged images for a specified container package:</p>
<pre><code class="lang-auto">      - uses: actions/github-script@v3
        with:
          github-token: ${{ secrets.DELETE_PACKAGES_TOKEN }}
          script: |
            const response = await github.request("GET /${{ env.OWNER }}/packages/container/${{ env.PACKAGE_NAME }}/versions",
              { per_page: ${{ env.PER_PAGE }}
            });
            for(version of response.data) {
                if (version.metadata.container.tags.length == 0) {
                    console.log("delete " + version.id)
                    const deleteResponse = await github.request("DELETE /${{ env.OWNER }}/packages/container/${{ env.PACKAGE_NAME }}/versions/" + version.id, { });
                    console.log("status " + deleteResponse.status)
                }
            }
        env:
          OWNER: user # or orgs/&lt;org name&gt;
          PACKAGE_NAME: &lt;package name&gt;
          PER_PAGE: 100
</code></pre>
<p><code>OWNER</code> should be either a user name or <code>orgs/ORG_NAME</code>.<br>
<code>DELETE_PACKAGES_TOKEN</code> is a PAT with the <code>delete:packages</code> and <code>write:packages</code> scopes.</p>
<p>If anyone tries this, please let me know if you have any problems running the script!</p>

**Reply by lakinduakash** [reply]
@jcansdale
> Hi [@airtower-luna](/airtower-luna),
> 
> > I’ve written the kind of cleanup script I was thinking of here using the new [packages delete/restore API ](https://github.blog/changelog/2021-02-17-packages-delete-and-restore-for-all-packages/). 🙂
> 
> I had a similar thought. 😃
> 
> Here’s a GitHub Action script that can be used to delete untagged images for a specified container package:
> 
> ```
>       - uses: actions/github-script@v3
>         with:
>           github-token: ${{ secrets.DELETE_PACKAGES_TOKEN }}
>           script: |
>             const response = await github.request("GET /${{ env.OWNER }}/packages/container/${{ env.PACKAGE_NAME }}/versions",
>               { per_page: ${{ env.PER_PAGE }}
>             });
>             for(version of response.data) {
>                 if (version.metadata.container.tags.length == 0) {
>                     console.log("delete " + version.id)
>                     const deleteResponse = await github.request("DELETE /${{ env.OWNER }}/packages/container/${{ env.PACKAGE_NAME }}/versions/" + version.id, { });
>                     console.log("status " + deleteResponse.status)
>                 }
>             }
>         env:
>           OWNER: user # or orgs/<org name>
>           PACKAGE_NAME: <package name>
>           PER_PAGE: 100
> ```
> 
> `OWNER` should be either a user name or `orgs/ORG_NAME`. `DELETE_PACKAGES_TOKEN` is a PAT with the `delete:packages` and `write:packages` scopes.
> 
> If anyone tries this, please let me know if you have any problems running the script!

OH, thanks! I tried every old package deletion GitHub action packages. None worked. This worked!

**Reply by lincior** [reply]
> `OWNER` should be either a user name or `orgs/ORG_NAME`. `DELETE_PACKAGES_TOKEN` is a PAT with the `delete:packages` and `write:packages` scopes.

@jcansdale thanks for the workflow, it's great. By chance have you tried to avoid using a PAT and using instead [jobs-level permissions](https://docs.github.com/en/actions/using-jobs/assigning-permissions-to-jobs)?

**Reply by mering** [reply]
I would be interested in a solution using jobs-level permissions as well!

**Comment by ain-soph** [comment]
<p>Just want to clarify that <code>OWNER</code> needn’t change to be your username, just keeping it to be <code>"user"</code>. I’ve been stuck on that for quite some time.</p>

**Comment by jcansdale** [comment]
<aside class="quote no-group" data-username="ain-soph" data-post="13" data-topic="132977" data-full="true">
<div class="title">
<div class="quote-controls"></div>
 ain-soph:</div>
<blockquote>
<p>Just want to clarify that <code>OWNER</code> needn’t change to be your username, just keeping it to be <code>"user"</code> . I’ve been stuck on that for quite some time.</p>
</blockquote>
</aside>
<p>Sorry about that! I forgot and fell into this trap as well. :blush:</p>

**Reply by pkatul** [reply]
First of all, this script was really helpful, 
For my I had to use orgs/ORG_NAME to make it working.

**Comment by ghost** [comment]
<p>Hi, there.</p>
<p>Thank you for your script. It is working as intended.</p>
<p>However, if you build a multi-arch image with <span class="mention">@latest</span> tag looks like only one arch will be tagged which means the container of rest arch will be deleted. Could you tell me you have a solution for this case please?</p>

**Reply by corinz** [reply]
@JonSun30 Did you ever figure this out?

**Reply by ghost** [reply]
Initially docker images supported only one architecture, so the structure to describe the image, named "manifest", supports only that. When multi-arch docker images were introduced, a "manifest list" structure was created, that contains the supported architecture names and references to manifests for each architecture. So if you create a multi-arch image supporting linux/amd64 and linux/arm64, three structures (with corresponding digests) are created: A manifest list, a manifest for the linux/amd64 architecture and another one for the linux/arm64 architecture.

Tags for single-arch images are tagging the manifest, for multi-arch images the manifest list gets tagged.

As an example look at the [version page of turboscan-docker-images/mysql](https://github.com/orgs/github/packages/container/turboscan-docker-images%2Fmysql/versions). You can't distingush if an item is a manifest or a manifest lists, even using the GitHub API won't help. The offical way is to load the contents of the digests. In the web interface you can follow the digest (or tag) link and you will notice a different page for the first item than that for the other ones. The first item is a manifest list, the other ones are manifests. In the manifest list of the first item in the "OS / Arch" tab the digests of the second and third item are referenced, so these three items belong to one multi-arch image. The forth and fifth item are not referenced by the manifest list, so they are two single-arch images.

Deleting manifests, that are referenced by a manifest list should be avoided. My solution (more a workaround) is to look at the creation dates. Multi-arch images are creating the manifests and the manifest list at the same time, so their creation date is within a short time span, for example here the information for the mysql images using the GitHub API:

```
$ curl -s -H "Authorization: token $TOKEN" 'https://api.github.com/orgs/github/packages/container/turboscan-docker-images%2Fmysql/versions' | jq -r '.[] | "\(.name) \(.created_at)"'
sha256:cdc0473e945049c9c32ccb10299c955d5be398a6a2a00ef2c060c538d5cd72d8 2022-03-17T13:18:15Z
sha256:c98b4ff5d7f4b81df412008da476e5fc6ea5fff71e7947a3ba80d2028362b74d 2022-03-17T13:18:14Z
sha256:ea49aee4acb5d1191114abfb7ffae73c785f7e37c9ba6f03a808e53ba154cb2f 2022-03-17T13:18:13Z
sha256:f0910a67381037cac57bfcef16645709cc113846ad826ba08da01716973d4a28 2022-03-17T13:17:32Z
sha256:11a5376741b847fb608ad362228b991fcaadd42af0371e2c7a25f4ad5ba94ab2 2022-03-17T13:12:35Z
```

The first three digests are created within a short time frame, they should be handled as a single entity. So either delete all three of them or none. The fourth and fifth digest don't belong to the manifest list and could be deleted at any time.
